<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>sankey_amr</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="sankey_amr_files/libs/clipboard/clipboard.min.js"></script>
<script src="sankey_amr_files/libs/quarto-html/quarto.js"></script>
<script src="sankey_amr_files/libs/quarto-html/popper.min.js"></script>
<script src="sankey_amr_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="sankey_amr_files/libs/quarto-html/anchor.min.js"></script>
<link href="sankey_amr_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="sankey_amr_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="sankey_amr_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="sankey_amr_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="sankey_amr_files/libs/bootstrap/bootstrap-1bc8a17f135ab3d594c857e9f48e611b.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="sankey_amr_files/libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>

<script src="sankey_amr_files/libs/d3-4.9.1/d3.min.js"></script>

<script src="sankey_amr_files/libs/sankey-1/sankey.js"></script>

<script src="sankey_amr_files/libs/sankeyNetwork-binding-0.4.1/sankeyNetwork.js"></script>



</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">




<section id="sankey-plot-to-visualise-e.-coli-resistance-genes" class="level2">
<h2 class="anchored" data-anchor-id="sankey-plot-to-visualise-e.-coli-resistance-genes">Sankey plot to visualise <em>E. coli</em> resistance genes</h2>
<p>Here are the steps I used to generate my Sankey diagram.</p>
<section id="packages" class="level3">
<h3 class="anchored" data-anchor-id="packages">Packages</h3>
<p>If you haven’t done so already install the packages networkD3, tidyverse, htmlwidgets, manipulateWidget and webshot2.</p>
<p>Load the libraries</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(networkD3)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(htmlwidgets)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(manipulateWidget)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(webshot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="read-in-and-prepare-data" class="level3">
<h3 class="anchored" data-anchor-id="read-in-and-prepare-data">Read in and prepare data</h3>
<p>The following code was adapted from <a href="https://r-graph-gallery.com/321-introduction-to-interactive-sankey-diagram-2.html" class="uri">https://r-graph-gallery.com/321-introduction-to-interactive-sankey-diagram-2.html</a> (and a little bit of help from chatgpt when I had trouble setting up the groups). Two dataframes are needed to generate a Sankey diagram. One for the links and the other for the nodes. In the r graph gallery example the links dataframe was generated using the data.frame() function. Initially I tried this, but then it became too messy when I was adding multiple antibiotic classes so I read in a csv file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>links <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/resistance_mechanisms.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 74 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): source, target
dbl (1): value

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>My csv file contains three columns: source, target, value. Source is the origin of each data link. Given, my Sankey diagram has three columns, source is both the name of the of the antibiotic classes and the mechanisms. Target is the endpoint of each data link. This contains the mechanisms and resistance genes.Value is the relative thickness of each link.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#First six lines of my data frame</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(links)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  source           target                 value
  &lt;chr&gt;            &lt;chr&gt;                  &lt;dbl&gt;
1 Aminoglycosides  Enzymatic inactivation    14
2 Aminoglycosides  Target alteration          8
3 Aminoglycosides  Efflux pump                2
4 Fluoroquinolones Enzymatic inactivation     2
5 Fluoroquinolones Target alteration          8
6 Fluoroquinolones Target protection          2</code></pre>
</div>
</div>
<p>The ` in some of my gene names is read in as ? so first I replace the ? with ′</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>links<span class="sc">$</span>target <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(links<span class="sc">$</span>target, <span class="st">"</span><span class="sc">\\</span><span class="st">?"</span>, <span class="st">"′"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next I created a one column data frame for all the nodes.The column name is labelled name and the observations are all the unique names from source and target.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">as.character</span>(links<span class="sc">$</span>source), <span class="fu">as.character</span>(links<span class="sc">$</span>target))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A second column was added to the nodes dataframe called group. This will be used to colour the Sankey diagram links by group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Grouping</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>nodes<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">case_when</span>(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Aminoglycosides"</span> <span class="sc">~</span> <span class="st">"Aminoglycosides"</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Fluoroquinolones"</span> <span class="sc">~</span> <span class="st">"Fluoroquinolones"</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Tetracyclines"</span> <span class="sc">~</span> <span class="st">"Tetracyclines"</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Phenicols"</span> <span class="sc">~</span> <span class="st">"Phenicols"</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Sulfonamides &amp; Trimethoprim"</span> <span class="sc">~</span> <span class="st">"Sulfonamides"</span>,</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Nitrofurans"</span> <span class="sc">~</span> <span class="st">"Nitrofurans"</span>,</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Macrolides"</span> <span class="sc">~</span> <span class="st">"Macrolides"</span>,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"aac(3′)"</span>, <span class="st">"aadA"</span>, <span class="st">"aadB"</span>, <span class="st">"aadD"</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"aph(3′)"</span>, <span class="st">"aphA15"</span>, <span class="st">"rrsH*"</span>, <span class="st">"rsmG*"</span>) <span class="sc">~</span> <span class="st">"Aminoglycosides"</span>,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"gyrA*"</span>, <span class="st">"gyrB*"</span>, <span class="st">"parC*"</span>, <span class="st">"parE*"</span>, <span class="st">"qep"</span>, <span class="st">"qnr"</span>, <span class="st">"oqxAB"</span>) <span class="sc">~</span> <span class="st">"Fluoroquinolones"</span>,</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"tetX"</span>, <span class="st">"tetM"</span>, <span class="st">"tetW"</span>, <span class="st">"tetA"</span>, <span class="st">"tetB"</span>, <span class="st">"tetC"</span>, <span class="st">"tetD"</span>) <span class="sc">~</span> <span class="st">"Tetracyclines"</span>,</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"catB"</span>, <span class="st">"catII"</span>, <span class="st">"cmlA"</span>, <span class="st">"floR"</span>) <span class="sc">~</span> <span class="st">"Phenicols"</span>,</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"dfr"</span>, <span class="st">"folP*"</span>, <span class="st">"sul1"</span>, <span class="st">"sul2"</span>, <span class="st">"sul3"</span>, <span class="st">"sul4"</span>) <span class="sc">~</span> <span class="st">"Sulfonamides"</span>,</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"nfsA*"</span>, <span class="st">"nfsB*"</span>, <span class="st">"ahpF*"</span>, <span class="st">"ribB*"</span>, <span class="st">"ribE*"</span>) <span class="sc">~</span> <span class="st">"Nitrofurans"</span>,</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"ermA"</span>, <span class="st">"ermB"</span>, <span class="st">"ermC"</span>, <span class="st">"ereA"</span>, <span class="st">"mphA"</span>, <span class="st">"mphB"</span>, <span class="st">"mefA"</span>, <span class="st">"mefB"</span>, <span class="st">"msrA"</span>, <span class="st">"msrD"</span>) <span class="sc">~</span> <span class="st">"Macrolides"</span>,</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>  <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Other"</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Two additional columns were added, which contain an index for both the source and the target. The match function is used to match the position of the values that are being matched. As I understand it networkD3 uses an index system starting from 0, whereas base R starts from 1, hence the -1 is needed to generate an index system suitable for the function networkD3::sankeyNetwork().</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ID mapping</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>links<span class="sc">$</span>IDsource <span class="ot">&lt;-</span> <span class="fu">match</span>(links<span class="sc">$</span>source, nodes<span class="sc">$</span>name) <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>links<span class="sc">$</span>IDtarget <span class="ot">&lt;-</span> <span class="fu">match</span>(links<span class="sc">$</span>target, nodes<span class="sc">$</span>name) <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>links</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 74 × 5
   source           target                 value IDsource IDtarget
   &lt;chr&gt;            &lt;chr&gt;                  &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
 1 Aminoglycosides  Enzymatic inactivation    14        0        7
 2 Aminoglycosides  Target alteration          8        0        8
 3 Aminoglycosides  Efflux pump                2        0        9
 4 Fluoroquinolones Enzymatic inactivation     2        1        7
 5 Fluoroquinolones Target alteration          8        1        8
 6 Fluoroquinolones Target protection          2        1       10
 7 Fluoroquinolones Efflux pump                6        1        9
 8 Tetracyclines    Enzymatic inactivation     2        2        7
 9 Tetracyclines    Target alteration          4        2        8
10 Tetracyclines    Target protection          4        2       10
# ℹ 64 more rows</code></pre>
</div>
</div>
<p>Similar to the nodes dataframe a group column was also added to the links dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Generate vector containing the group names for all the antibiotic classes</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>class <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Aminoglycosides"</span>, <span class="st">"Fluoroquinolones"</span>, <span class="st">"Tetracyclines"</span>, <span class="st">"Phenicols"</span>, <span class="st">"Sulfonamides &amp; Trimethoprim"</span>, <span class="st">"Nitrofurans"</span>, <span class="st">"Macrolides"</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Generate vector containing the group names for all the mechanisms</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>mechanisms <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Enzymatic inactivation"</span>, <span class="st">"Target alteration"</span>, <span class="st">"Target protection"</span>, <span class="st">"Efflux pump"</span>, <span class="st">"Target replacement"</span>, <span class="st">"Reduced prodrug activation"</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign groups</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>links<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">case_when</span>(</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  links<span class="sc">$</span>source <span class="sc">%in%</span> class <span class="sc">~</span> nodes<span class="sc">$</span>group[<span class="fu">match</span>(links<span class="sc">$</span>source, nodes<span class="sc">$</span>name)],</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  links<span class="sc">$</span>source <span class="sc">%in%</span> mechanisms <span class="sc">~</span> nodes<span class="sc">$</span>group[<span class="fu">match</span>(links<span class="sc">$</span>target, nodes<span class="sc">$</span>name)],</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Other"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lastly I generated my colour scale. Here (as I understand it) because networkD3 uses Javascript code the colour scale also has to be written in Javascript. ‘.domain’ defines each group that I want to apply each colour to.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>my_color <span class="ot">&lt;-</span> <span class="st">'d3.scaleOrdinal()</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="st">  .domain(["Aminoglycosides", "Fluoroquinolones", "Tetracyclines", </span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="st">           "Phenicols", "Sulfonamides", "Nitrofurans", "Macrolides", "Other"])</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="st">  .range(["#90CAF9", "#A5D6A7", "#D1C4E9", "#ffffb3", "#fb8072", "#fdb462", "#F8BBD0FF", "#E0E0E0"]);'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sankey-diagram" class="level3">
<h3 class="anchored" data-anchor-id="sankey-diagram">Sankey diagram</h3>
<p>Now for the fun part generating the Sankey diagram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sankey plot</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> <span class="fu">sankeyNetwork</span>(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Links =</span> links, <span class="at">Nodes =</span> nodes,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Source =</span> <span class="st">"IDsource"</span>, <span class="at">Target =</span> <span class="st">"IDtarget"</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Value =</span> <span class="st">"value"</span>, <span class="at">NodeID =</span> <span class="st">"name"</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">NodeGroup =</span> <span class="st">"group"</span>, <span class="at">LinkGroup =</span> <span class="st">"group"</span>,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">colourScale =</span> my_color,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">900</span>, <span class="at">height =</span> <span class="dv">900</span>,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">nodeWidth =</span> <span class="dv">60</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Links is a tbl_df. Converting to a plain data frame.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>q</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="htmlwidget-0fc3d4874d5e9dc15e34" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-0fc3d4874d5e9dc15e34">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Reduced prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":7,"fontFamily":null,"nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":true}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>There are a few things I don’t like. The font size and type, the width of the nodes and having the resistance genes on top of the links.</p>
<p>Here I’ve adjusted these four things within the sankeyNetwork() function using the arguments fontSize, fontFamily, nodeWidth and sinksRight.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sankey plot</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">sankeyNetwork</span>(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Links =</span> links, <span class="at">Nodes =</span> nodes,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Source =</span> <span class="st">"IDsource"</span>, <span class="at">Target =</span> <span class="st">"IDtarget"</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Value =</span> <span class="st">"value"</span>, <span class="at">NodeID =</span> <span class="st">"name"</span>,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">NodeGroup =</span> <span class="st">"group"</span>, <span class="at">LinkGroup =</span> <span class="st">"group"</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">colourScale =</span> my_color,</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">fontSize =</span> <span class="dv">12</span>, <span class="at">fontFamily =</span> <span class="st">"Verdana"</span>, <span class="at">nodeWidth =</span> <span class="dv">60</span>, <span class="at">sinksRight =</span> <span class="cn">FALSE</span>, <span class="at">width =</span> <span class="dv">900</span>, <span class="at">height =</span> <span class="dv">900</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Links is a tbl_df. Converting to a plain data frame.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="htmlwidget-1c0256f7ca3bf81656c2" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-1c0256f7ca3bf81656c2">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Reduced prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":12,"fontFamily":"Verdana","nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>It’s still not quite what I’d like. To my knowledge that is pretty much all that can be adjusted within the SankeyNetwork() function. However, because the Sankey Network is an htmlwidget (you can check this using the class function) you can use the htmlwidge package to append labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sankeyNetwork" "htmlwidget"   </code></pre>
</div>
</div>
</section>
<section id="using-htmlwidgets-to-adjust-the-sankey-diagram" class="level3">
<h3 class="anchored" data-anchor-id="using-htmlwidgets-to-adjust-the-sankey-diagram">Using htmlwidgets to adjust the Sankey diagram</h3>
<p>The code below was adapted from the tutorial https://www.geeksforgeeks.org/r-language/how-to-add-axis-labels-using-networkd3-in-r/ and the response on this stack overflow post https://stackoverflow.com/questions/50132459/how-to-add-title-to-a-networkd3-visualisation-when-saving-as-a-web-page</p>
<p>Here I’ve added a title to my figure and a footnote.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#prepend title</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> htmlwidgets<span class="sc">::</span><span class="fu">prependContent</span>(p, htmltools<span class="sc">::</span>tags<span class="sc">$</span><span class="fu">h1</span>(htmltools<span class="sc">::</span><span class="fu">HTML</span>(<span class="st">"Resistance genes in &lt;em&gt;Escherichia coli&lt;/em&gt;"</span>),</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">style =</span> <span class="st">"text-align:center; color:#1b2422; font-size:18px; font-family:Verdana;"</span>))</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>p1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<h1 style="text-align:center; color:#1b2422; font-size:18px; font-family:Verdana;">Resistance genes in <em>Escherichia coli</em></h1>
<div id="htmlwidget-f77b783f23039690868a" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-f77b783f23039690868a">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Reduced prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":12,"fontFamily":"Verdana","nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">#append footnote</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> htmlwidgets<span class="sc">::</span><span class="fu">appendContent</span>(p1, htmltools<span class="sc">::</span>tags<span class="sc">$</span><span class="fu">p</span>(<span class="st">"* Genes with point mutations"</span>, <span class="at">style =</span> <span class="st">"text-align:center; color:#666; font-size:12px; font-family:Verdana;"</span>))</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<h1 style="text-align:center; color:#1b2422; font-size:18px; font-family:Verdana;">Resistance genes in <em>Escherichia coli</em></h1>
<div id="htmlwidget-82374ba9156384f4f5ad" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<p style="text-align:center; color:#666; font-size:12px; font-family:Verdana;">* Genes with point mutations</p>
<script type="application/json" data-for="htmlwidget-82374ba9156384f4f5ad">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Reduced prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":12,"fontFamily":"Verdana","nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>I also wanted to change the font style to bold for antibiotic classes and italics for the resistance genes. I got a bit of help from chatgpt for this piece of code.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#The following function was generated using chapgpt</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> htmlwidgets<span class="sc">::</span><span class="fu">onRender</span>(p2, <span class="st">'</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="st">  function(el, x) {</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="st">    const boldNames = [</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="st">      "Aminoglycosides", "Fluoroquinolones", "Tetracyclines",</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="st">      "Phenicols", "Sulfonamides &amp; Trimethoprim", "Nitrofurans",</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="st">      "Enzymatic inactivation", "Target alteration",</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="st">      "Target protection", "Efflux pump", "Target replacement",</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="st">      "Reduced prodrug activation", "Macrolides"</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="st">    ];</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="st">    const italicSubstrings = [</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="st">      "aac", "aad", "aph", "rr", "rsm", "qep", "qnr", "tet", "cat",</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a><span class="st">      "gyr", "par", "mdf", "fol", "dfr", "nfs", "cml", "flo", "oqx", "sul",</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a><span class="st">      "erm", "ereA", "mph", "mef", "msr", "ahpF", "rib"</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="st">    ];</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a><span class="st">    d3.select(el)</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="st">      .selectAll(".node text")</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a><span class="st">      .style("font-weight", d =&gt; boldNames.includes(d.name) ? "bold" : "normal")</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a><span class="st">      .style("font-style", d =&gt;</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a><span class="st">        italicSubstrings.some(sub =&gt; d.name.includes(sub)) ? "italic" : "normal"</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a><span class="st">      );</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a><span class="st">  }</span></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a><span class="st">'</span>)</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>p3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<h1 style="text-align:center; color:#1b2422; font-size:18px; font-family:Verdana;">Resistance genes in <em>Escherichia coli</em></h1>
<div id="htmlwidget-0bbfda076deee9ea957e" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<p style="text-align:center; color:#666; font-size:12px; font-family:Verdana;">* Genes with point mutations</p>
<script type="application/json" data-for="htmlwidget-0bbfda076deee9ea957e">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Reduced prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":12,"fontFamily":"Verdana","nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":false}},"evals":[],"jsHooks":{"render":[{"code":"\n  function(el, x) {\n    const boldNames = [\n      \"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\",\n      \"Phenicols\", \"Sulfonamides & Trimethoprim\", \"Nitrofurans\",\n      \"Enzymatic inactivation\", \"Target alteration\",\n      \"Target protection\", \"Efflux pump\", \"Target replacement\",\n      \"Reduced prodrug activation\", \"Macrolides\"\n    ];\n\n    const italicSubstrings = [\n      \"aac\", \"aad\", \"aph\", \"rr\", \"rsm\", \"qep\", \"qnr\", \"tet\", \"cat\",\n      \"gyr\", \"par\", \"mdf\", \"fol\", \"dfr\", \"nfs\", \"cml\", \"flo\", \"oqx\", \"sul\",\n      \"erm\", \"ereA\", \"mph\", \"mef\", \"msr\", \"ahpF\", \"rib\"\n    ];\n\n    d3.select(el)\n      .selectAll(\".node text\")\n      .style(\"font-weight\", d => boldNames.includes(d.name) ? \"bold\" : \"normal\")\n      .style(\"font-style\", d =>\n        italicSubstrings.some(sub => d.name.includes(sub)) ? \"italic\" : \"normal\"\n      );\n  }\n","data":null}]}}</script>
</div>
</div>
</section>
</section>
<section id="saving-the-sankey-diagram" class="level2">
<h2 class="anchored" data-anchor-id="saving-the-sankey-diagram">Saving the Sankey diagram</h2>
<p>If you are trying this code or similar and want to save your plot as a png you will not be able to use ggsave Because the Sankey Network is an htmlwidget. First save your plot as an html and then take a screenshot using webshot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>htmlwidgets<span class="sc">::</span><span class="fu">saveWidget</span>(p3, <span class="st">"output/sankey.html"</span>, <span class="at">selfcontained =</span> <span class="cn">TRUE</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">webshot</span>(<span class="st">"output/sankey.html"</span>, <span class="at">file =</span> <span class="st">"output/sankey_arg.png"</span>, <span class="at">vwidth =</span> <span class="dv">900</span>, <span class="at">vheight =</span> <span class="dv">900</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>One thing I haven’t been able to work out is how remove the white space on either side and align the footer to the gene name column, when I save the plot as a png. If anyone reading this has any tips please email me or DM on LinkedIn.</p>
<p>Lastly here are some references I used for generating the resistance gene list:</p>

</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-kerek2025" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Kerek, Á. <em>et al.</em> Antibiotic resistance genes in <span><em>Escherichia coli</em></span> <span></span> literature review. <em>Critical Reviews in Microbiology</em> 1–35 (2025) doi:<a href="https://doi.org/10.1080/1040841X.2025.2492156">10.1080/1040841X.2025.2492156</a>.</div>
</div>
<div id="ref-gomi2025" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Gomi, R. &amp; Adachi, F. <a href="https://doi.org/10.1007/s00248-025-02502-7">Quinolone Resistance Genes qnr, aac(6<span>’</span>)-Ib-cr, oqxAB, and qepA in Environmental Escherichia coli: Insights into Their Genetic Contexts from Comparative Genomics</a>. <em>Microbial Ecology</em> <strong>88</strong>, 6 (2025).</div>
</div>
<div id="ref-le2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Le, V. V. H. &amp; Rakonjac, J. <a href="https://doi.org/10.1371/journal.ppat.1009663">Nitrofurans: Revival of an <span>“</span>old<span>”</span> drug class in the fight against antibiotic resistance</a>. <em>PLOS Pathogens</em> <strong>17</strong>, e1009663 (2021).</div>
</div>
<div id="ref-nguyen2009" class="csl-entry" role="listitem">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Nguyen, M. C. P. <em>et al.</em> <a href="https://doi.org/10.3201/eid1510.090696"><span><em>Escherichia coli</em></span> as reservoir for macrolide resistance genes</a>. <em>Emerging Infectious Diseases</em> <strong>15</strong>, 1648–1650 (2009).</div>
</div>
<div id="ref-vedantam1998" class="csl-entry" role="listitem">
<div class="csl-left-margin">5. </div><div class="csl-right-inline">Vedantam, G., Guay, G. G., Austria, N. E., Doktor, S. Z. &amp; Nichols, B. P. <a href="https://doi.org/10.1128/AAC.42.1.88">Characterization of Mutations Contributing to Sulfathiazole Resistance in <span><em>Escherichia coli</em></span></a>. <em>Antimicrobial Agents and Chemotherapy</em> <strong>42</strong>, 88–93 (1998).</div>
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>