<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Generating and polishing Sankey diagrams in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d4be639c637f3db3c684c66cefad7e0c.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>

<script src="site_libs/d3-4.9.1/d3.min.js"></script>

<script src="site_libs/sankey-1/sankey.js"></script>

<script src="site_libs/sankeyNetwork-binding-0.4.1/sankeyNetwork.js"></script>


  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="fullcontent">

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"><h1 style="text-align:center; color:#1b2422; font-size:18px; font-family:Verdana;" class="title display-7">Resistance genes in <em>Escherichia coli</em></h1></header>




<section id="generating-and-polishing-sankey-diagrams-in-r" class="level1">
<h1>Generating and polishing Sankey diagrams in R</h1>
<p>Sara Burgess, 14 November 2025</p>
<p>Sankey diagrams are used to illustrate flows from one item (node) to another. There are several R packages used to generate Sankey diagrams. This blog will demonstrate how to use the <code>networkD3</code> and <code>ggsankey</code> packages.</p>
<section id="data" class="level2">
<h2 class="anchored" data-anchor-id="data">Data</h2>
<p>In these examples data was manually sourced from publications (as listed at the bottom of this blog) to visualise the different resistance genes and mechanisms that the bacterium <em>Escherichia coli</em> can use to become resistant to seven classes of antibiotics. The first Sankey diagram covers <em>E. coli</em> resistance to six classes of antibiotics (aminoglycosides, fluoroquinolones, tetracyclines, phenicols, sulfonamides and trimethoprim, nitrofurans). The second Sankey diagram shows how different beta-lactamase enzyme types confer resistance to different sub-classes of beta-lactams.</p>
</section>
<section id="generating-a-sankey-diagram-using-networkd3" class="level2">
<h2 class="anchored" data-anchor-id="generating-a-sankey-diagram-using-networkd3">Generating a Sankey diagram using networkD3</h2>
<p>The package networkD3 can be used for generating a variety of interactive network (as described in this geeksforgeeks <a href="https://www.geeksforgeeks.org/r-language/networkd3-package-in-r/">blog</a>.</p>
<section id="packages" class="level3">
<h3 class="anchored" data-anchor-id="packages">Packages</h3>
<p>If you haven’t done so already install the packages <code>networkD3</code>, <code>tidyverse</code>, <code>htmlwidgets</code>, <code>manipulateWidget</code> and <code>webshot2</code>.</p>
<p>Load the libraries</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(networkD3)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(htmlwidgets)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(manipulateWidget)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(webshot2)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(paletteer)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="read-in-and-prepare-data" class="level3">
<h3 class="anchored" data-anchor-id="read-in-and-prepare-data">Read in and prepare data</h3>
<p>The following code was adapted from the <a href="https://r-graph-gallery.com/321-introduction-to-interactive-sankey-diagram-2.html">r-graph-gallery</a> . Two data frames are needed to generate a Sankey diagram. One for the links and the other for the nodes. First read in your data for the links data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>links <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/resistance_mechanisms.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 75 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): source, target
dbl (1): value

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>My csv file contains three columns: source, target, value. Source is the origin of each data link. Given, my Sankey diagram has three columns, source contains both the names of the antibiotic classes and the mechanisms. Target is the endpoint of each data link. This contains the mechanisms and resistance genes. Value is the thickness of each link.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#First six lines of my data frame</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(links)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  source           target                 value
  &lt;chr&gt;            &lt;chr&gt;                  &lt;dbl&gt;
1 Aminoglycosides  Enzymatic inactivation    14
2 Aminoglycosides  Target alteration          8
3 Aminoglycosides  Efflux pump                2
4 Fluoroquinolones Enzymatic inactivation     2
5 Fluoroquinolones Target alteration          8
6 Fluoroquinolones Target protection          2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Last six lines of my data frame</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(links)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  source      target value
  &lt;chr&gt;       &lt;chr&gt;  &lt;dbl&gt;
1 Efflux pump mphA       2
2 Efflux pump mphB       2
3 Efflux pump mefA       2
4 Efflux pump mefB       2
5 Efflux pump msrA       2
6 Efflux pump msrD       2</code></pre>
</div>
</div>
<p>The <code>'</code> in some of my gene names is read in as <code>?</code> so first I replaced the <code>?</code> with <code>′.</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>links<span class="sc">$</span>target <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(links<span class="sc">$</span>target, <span class="st">"</span><span class="sc">\\</span><span class="st">?"</span>, <span class="st">"′"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next I created a one column data frame for all the nodes. The column name is labelled name and the observations are all the unique names from source and target.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">as.character</span>(links<span class="sc">$</span>source), <span class="fu">as.character</span>(links<span class="sc">$</span>target))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A second column was added to the nodes data frame called group. This will be used to colour the Sankey diagram links by group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Grouping</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>nodes<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">case_when</span>(</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Aminoglycosides"</span> <span class="sc">~</span> <span class="st">"Aminoglycosides"</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Fluoroquinolones"</span> <span class="sc">~</span> <span class="st">"Fluoroquinolones"</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Tetracyclines"</span> <span class="sc">~</span> <span class="st">"Tetracyclines"</span>,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Phenicols"</span> <span class="sc">~</span> <span class="st">"Phenicols"</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Sulfonamides &amp; Trimethoprim"</span> <span class="sc">~</span> <span class="st">"Sulfonamides"</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Nitrofurans"</span> <span class="sc">~</span> <span class="st">"Nitrofurans"</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"Macrolides"</span> <span class="sc">~</span> <span class="st">"Macrolides"</span>,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"aac(3′)"</span>, <span class="st">"aadA"</span>, <span class="st">"aadB"</span>, <span class="st">"aadD"</span>,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"aph(3′)"</span>, <span class="st">"aphA15"</span>, <span class="st">"rrsH*"</span>, <span class="st">"rsmG*"</span>) <span class="sc">~</span> <span class="st">"Aminoglycosides"</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"gyrA*"</span>, <span class="st">"gyrB*"</span>, <span class="st">"parC*"</span>, <span class="st">"parE*"</span>, <span class="st">"qep"</span>, <span class="st">"qnr"</span>, <span class="st">"oqxAB"</span>) <span class="sc">~</span> <span class="st">"Fluoroquinolones"</span>,</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"tetX"</span>, <span class="st">"tetM"</span>, <span class="st">"tetW"</span>, <span class="st">"tetA"</span>, <span class="st">"tetB"</span>, <span class="st">"tetC"</span>, <span class="st">"tetD"</span>) <span class="sc">~</span> <span class="st">"Tetracyclines"</span>,</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"catB"</span>, <span class="st">"catII"</span>, <span class="st">"cmlA"</span>, <span class="st">"floR"</span>) <span class="sc">~</span> <span class="st">"Phenicols"</span>,</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"dfr"</span>, <span class="st">"folP*"</span>, <span class="st">"sul1"</span>, <span class="st">"sul2"</span>, <span class="st">"sul3"</span>, <span class="st">"sul4"</span>) <span class="sc">~</span> <span class="st">"Sulfonamides"</span>,</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"nfsA*"</span>, <span class="st">"nfsB*"</span>, <span class="st">"ahpF*"</span>, <span class="st">"ribB*"</span>, <span class="st">"ribE*"</span>) <span class="sc">~</span> <span class="st">"Nitrofurans"</span>,</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>  nodes<span class="sc">$</span>name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"ermA"</span>, <span class="st">"ermB"</span>, <span class="st">"ermC"</span>, <span class="st">"ereA"</span>, <span class="st">"mphA"</span>, <span class="st">"mphB"</span>, <span class="st">"mefA"</span>, <span class="st">"mefB"</span>, <span class="st">"msrA"</span>, <span class="st">"msrD"</span>) <span class="sc">~</span> <span class="st">"Macrolides"</span>,</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>  <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Other"</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Two additional columns were added, which contain an index for both the source and the target. The match function is used to match the position of the values that are being matched. As I understand it networkD3 uses an index system starting from 0, whereas base R starts from 1, hence the -1 is needed to generate an index system suitable for the function <code>networkD3::sankeyNetwork()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ID mapping</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>links<span class="sc">$</span>IDsource <span class="ot">&lt;-</span> <span class="fu">match</span>(links<span class="sc">$</span>source, nodes<span class="sc">$</span>name) <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>links<span class="sc">$</span>IDtarget <span class="ot">&lt;-</span> <span class="fu">match</span>(links<span class="sc">$</span>target, nodes<span class="sc">$</span>name) <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>links</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 75 × 5
   source           target                 value IDsource IDtarget
   &lt;chr&gt;            &lt;chr&gt;                  &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
 1 Aminoglycosides  Enzymatic inactivation    14        0        7
 2 Aminoglycosides  Target alteration          8        0        8
 3 Aminoglycosides  Efflux pump                2        0        9
 4 Fluoroquinolones Enzymatic inactivation     2        1        7
 5 Fluoroquinolones Target alteration          8        1        8
 6 Fluoroquinolones Target protection          2        1       10
 7 Fluoroquinolones Efflux pump                6        1        9
 8 Tetracyclines    Enzymatic inactivation     2        2        7
 9 Tetracyclines    Target alteration          4        2        8
10 Tetracyclines    Target protection          4        2       10
# ℹ 65 more rows</code></pre>
</div>
</div>
<p>Similar to the nodes data frame a group column was also added to the links data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Generate vector containing the group names for all the antibiotic classes</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>class <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Aminoglycosides"</span>, <span class="st">"Fluoroquinolones"</span>, <span class="st">"Tetracyclines"</span>, <span class="st">"Phenicols"</span>, <span class="st">"Sulfonamides &amp; Trimethoprim"</span>, <span class="st">"Nitrofurans"</span>, <span class="st">"Macrolides"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Generate vector containing the group names for all the mechanisms</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>mechanisms <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Enzymatic inactivation"</span>, <span class="st">"Target alteration"</span>, <span class="st">"Target protection"</span>, <span class="st">"Efflux pump"</span>, <span class="st">"Target replacement"</span>, <span class="st">"Reduced prodrug activation"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign groups</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>links<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">case_when</span>(</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  links<span class="sc">$</span>source <span class="sc">%in%</span> class <span class="sc">~</span> nodes<span class="sc">$</span>group[<span class="fu">match</span>(links<span class="sc">$</span>source, nodes<span class="sc">$</span>name)],</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  links<span class="sc">$</span>source <span class="sc">%in%</span> mechanisms <span class="sc">~</span> nodes<span class="sc">$</span>group[<span class="fu">match</span>(links<span class="sc">$</span>target, nodes<span class="sc">$</span>name)],</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Other"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lastly I generated my colour scale. Here because <code>networkD3</code> uses Javascript code the colour scale also has to be written in Javascript. <code>'.domain'</code> defines each group that I want to apply each colour to.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>my_color <span class="ot">&lt;-</span> <span class="st">'d3.scaleOrdinal()</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="st">  .domain(["Aminoglycosides", "Fluoroquinolones", "Tetracyclines", </span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="st">           "Phenicols", "Sulfonamides", "Nitrofurans", "Macrolides", "Other"])</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="st">  .range(["#90CAF9", "#A5D6A7", "#D1C4E9", "#ffffb3", "#fb8072", "#fdb462", "#F8BBD0FF", "#E0E0E0"]);'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sankey-diagram" class="level3">
<h3 class="anchored" data-anchor-id="sankey-diagram">Sankey diagram</h3>
<p>Now for the fun part generating the Sankey diagram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sankey plot</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> <span class="fu">sankeyNetwork</span>(</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Links =</span> links, <span class="at">Nodes =</span> nodes,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Source =</span> <span class="st">"IDsource"</span>, <span class="at">Target =</span> <span class="st">"IDtarget"</span>,</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Value =</span> <span class="st">"value"</span>, <span class="at">NodeID =</span> <span class="st">"name"</span>,</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">NodeGroup =</span> <span class="st">"group"</span>, <span class="at">LinkGroup =</span> <span class="st">"group"</span>,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">colourScale =</span> my_color,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">900</span>, <span class="at">height =</span> <span class="dv">900</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">nodeWidth =</span> <span class="dv">60</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>q</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="htmlwidget-7fb081b49f8c8361f0f6" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-7fb081b49f8c8361f0f6">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Other","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Other","Other","Other","Other","Other","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Poor prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","ydhC","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Other","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":7,"fontFamily":null,"nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":true}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>There are a few things I don’t like. The font size and type, as well as having the resistance genes on top of the links.</p>
<p>Here I’ve adjusted these three things within the <code>sankeyNetwork()</code> function using the arguments fontSize, fontFamily, and sinksRight.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sankey plot</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">sankeyNetwork</span>(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Links =</span> links, <span class="at">Nodes =</span> nodes,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Source =</span> <span class="st">"IDsource"</span>, <span class="at">Target =</span> <span class="st">"IDtarget"</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Value =</span> <span class="st">"value"</span>, <span class="at">NodeID =</span> <span class="st">"name"</span>,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">NodeGroup =</span> <span class="st">"group"</span>, <span class="at">LinkGroup =</span> <span class="st">"group"</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">colourScale =</span> my_color,</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">fontSize =</span> <span class="dv">12</span>, <span class="at">fontFamily =</span> <span class="st">"Verdana"</span>, <span class="at">nodeWidth =</span> <span class="dv">60</span>, <span class="at">sinksRight =</span> <span class="cn">FALSE</span>, <span class="at">width =</span> <span class="dv">900</span>, <span class="at">height =</span> <span class="dv">900</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="htmlwidget-d865d17ff34028fe4833" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-d865d17ff34028fe4833">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Other","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Other","Other","Other","Other","Other","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Poor prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","ydhC","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Other","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":12,"fontFamily":"Verdana","nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>It’s still not quite what I’d like. To my knowledge that is pretty much all that can be adjusted within the <code>SankeyNetwork()</code> function. However, because the Sankey diagram is an htmlwidget (you can check this using the class function) you can use the <code>htmlwidgets</code> package to append labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sankeyNetwork" "htmlwidget"   </code></pre>
</div>
</div>
</section>
<section id="using-htmlwidgets-to-adjust-the-sankey-diagram" class="level3">
<h3 class="anchored" data-anchor-id="using-htmlwidgets-to-adjust-the-sankey-diagram">Using htmlwidgets to adjust the Sankey diagram</h3>
<p>The code below was adapted from this <a href="https://www.geeksforgeeks.org/r-language/how-to-add-axis-labels-using-networkd3-in-r/" class="uri">tutorial</a> and the response on this <a href="https://stackoverflow.com/questions/50132459/how-to-add-title-to-a-networkd3-visualisation-when-saving-as-a-web-page">stack overflow post</a>.</p>
<p>First I added a title to my diagram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#prepend title</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> htmlwidgets<span class="sc">::</span><span class="fu">prependContent</span>(p, htmltools<span class="sc">::</span>tags<span class="sc">$</span><span class="fu">h1</span>(htmltools<span class="sc">::</span><span class="fu">HTML</span>(<span class="st">"Resistance genes in &lt;em&gt;Escherichia coli&lt;/em&gt;"</span>),</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">style =</span> <span class="st">"text-align:center; color:#1b2422; font-size:18px; font-family:Verdana;"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next a footnote was added.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#append footnote</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> htmlwidgets<span class="sc">::</span><span class="fu">appendContent</span>(p1, htmltools<span class="sc">::</span>tags<span class="sc">$</span><span class="fu">p</span>(<span class="st">"* Genes with point mutations"</span>, <span class="at">style =</span> <span class="st">"text-align:center; color:#666; font-size:12px; font-family:Verdana;"</span>))</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div id="htmlwidget-b417e1854efd0f1e4b49" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<p style="text-align:center; color:#666; font-size:12px; font-family:Verdana;">* Genes with point mutations</p>
<script type="application/json" data-for="htmlwidget-b417e1854efd0f1e4b49">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Other","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Other","Other","Other","Other","Other","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Poor prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","ydhC","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Other","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":12,"fontFamily":"Verdana","nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>I also changed the font style to bold for the antibiotic classes and italics for the resistance genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#The following function was generated using chapgpt</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> htmlwidgets<span class="sc">::</span><span class="fu">onRender</span>(p2, <span class="st">'</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="st">  function(el, x) {</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="st">    const boldNames = [</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="st">      "Aminoglycosides", "Fluoroquinolones", "Tetracyclines",</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="st">      "Phenicols", "Sulfonamides &amp; Trimethoprim", "Nitrofurans",</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="st">      "Enzymatic inactivation", "Target alteration",</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="st">      "Target protection", "Efflux pump", "Target replacement",</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="st">      "Reduced prodrug activation", "Macrolides"</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="st">    ];</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="st">    const italicSubstrings = [</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="st">      "aac", "aad", "aph", "rr", "rsm", "qep", "qnr", "tet", "cat",</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="st">      "gyr", "par", "mdf", "fol", "dfr", "nfs", "cml", "flo", "oqx", "sul",</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="st">      "erm", "ereA", "mph", "mef", "msr", "ahpF", "rib"</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="st">    ];</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a><span class="st">    d3.select(el)</span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a><span class="st">      .selectAll(".node text")</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a><span class="st">      .style("font-weight", d =&gt; boldNames.includes(d.name) ? "bold" : "normal")</span></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="st">      .style("font-style", d =&gt;</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="st">        italicSubstrings.some(sub =&gt; d.name.includes(sub)) ? "italic" : "normal"</span></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a><span class="st">      );</span></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a><span class="st">  }</span></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a><span class="st">'</span>)</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>p3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<h1 style="text-align:center; color:#1b2422; font-size:18px; font-family:Verdana;">Resistance genes in <em>Escherichia coli</em></h1>
<div id="htmlwidget-94487697e486c7b396ed" style="width:100%;height:650px;" class="sankeyNetwork html-widget"></div>
<p style="text-align:center; color:#666; font-size:12px; font-family:Verdana;">* Genes with point mutations</p>
<script type="application/json" data-for="htmlwidget-94487697e486c7b396ed">{"x":{"links":{"source":[0,0,0,1,1,1,1,2,2,2,2,3,3,4,4,4,5,6,6,7,7,7,7,7,7,7,8,8,8,8,9,7,8,8,8,8,10,9,9,9,7,8,8,10,10,9,9,9,9,7,7,9,9,9,8,8,11,11,11,11,12,12,12,12,12,7,7,7,7,9,9,9,9,9,9],"target":[7,8,9,7,8,10,9,7,8,10,9,7,9,8,11,9,12,7,7,13,14,15,16,17,18,19,20,21,22,23,24,14,25,26,27,28,29,30,31,24,32,21,20,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64],"value":[14,8,2,2,8,2,6,2,4,4,8,4,4,4,8,2,10,8,12,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],"group":["Aminoglycosides","Aminoglycosides","Aminoglycosides","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Macrolides","Macrolides","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Other","Tetracyclines","Other","Other","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Other","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Other","Other","Other","Other","Other","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"nodes":{"name":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides & Trimethoprim","Nitrofurans","Macrolides","Enzymatic inactivation","Target alteration","Efflux pump","Target protection","Target replacement","Poor prodrug activation","aac(3′)","aac(6′)","aadA","aadB","aadD","aph(3′)","aphA15","rrnB*","rrsB*","rrsH*","rsmG*","mdfA","gyrA*","gyrB*","parC*","parE*","qnr","qep","oqxAB","tetX","tetM","tetW","tetA","tetB","tetC","tetD","catB","catII","cmlA","floR","ydhC","folP*","dfr","sul1","sul2","sul3","sul4","nfsA*","nfsB*","ahpF*","ribB*","ribE*","ermA","ermB","ermC","ereA","mphA","mphB","mefA","mefB","msrA","msrD"],"group":["Aminoglycosides","Fluoroquinolones","Tetracyclines","Phenicols","Sulfonamides","Nitrofurans","Macrolides","Other","Other","Other","Other","Other","Other","Aminoglycosides","Other","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Aminoglycosides","Other","Other","Aminoglycosides","Aminoglycosides","Other","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Fluoroquinolones","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Tetracyclines","Phenicols","Phenicols","Phenicols","Phenicols","Other","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Sulfonamides","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Nitrofurans","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides","Macrolides"]},"options":{"NodeID":"name","NodeGroup":"group","LinkGroup":"group","colourScale":"d3.scaleOrdinal()\n  .domain([\"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\", \n           \"Phenicols\", \"Sulfonamides\", \"Nitrofurans\", \"Macrolides\", \"Other\"])\n  .range([\"#90CAF9\", \"#A5D6A7\", \"#D1C4E9\", \"#ffffb3\", \"#fb8072\", \"#fdb462\", \"#F8BBD0FF\", \"#E0E0E0\"]);","fontSize":12,"fontFamily":"Verdana","nodeWidth":60,"nodePadding":10,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":false}},"evals":[],"jsHooks":{"render":[{"code":"\n  function(el, x) {\n    const boldNames = [\n      \"Aminoglycosides\", \"Fluoroquinolones\", \"Tetracyclines\",\n      \"Phenicols\", \"Sulfonamides & Trimethoprim\", \"Nitrofurans\",\n      \"Enzymatic inactivation\", \"Target alteration\",\n      \"Target protection\", \"Efflux pump\", \"Target replacement\",\n      \"Reduced prodrug activation\", \"Macrolides\"\n    ];\n\n    const italicSubstrings = [\n      \"aac\", \"aad\", \"aph\", \"rr\", \"rsm\", \"qep\", \"qnr\", \"tet\", \"cat\",\n      \"gyr\", \"par\", \"mdf\", \"fol\", \"dfr\", \"nfs\", \"cml\", \"flo\", \"oqx\", \"sul\",\n      \"erm\", \"ereA\", \"mph\", \"mef\", \"msr\", \"ahpF\", \"rib\"\n    ];\n\n    d3.select(el)\n      .selectAll(\".node text\")\n      .style(\"font-weight\", d => boldNames.includes(d.name) ? \"bold\" : \"normal\")\n      .style(\"font-style\", d =>\n        italicSubstrings.some(sub => d.name.includes(sub)) ? \"italic\" : \"normal\"\n      );\n  }\n","data":null}]}}</script>
</div>
</div>
</section>
</section>
<section id="saving-the-sankey-diagram" class="level2">
<h2 class="anchored" data-anchor-id="saving-the-sankey-diagram">Saving the Sankey diagram</h2>
<p>If you are trying this code or similar and want to save your plot as a png you will not be able to use the <code>ggsave()</code> function because the Sankey diagram is an htmlwidget. First save your plot as an html and then take a screenshot using <code>webshot</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>htmlwidgets<span class="sc">::</span><span class="fu">saveWidget</span>(p3, <span class="st">"output/sankey.html"</span>, <span class="at">selfcontained =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">webshot</span>(<span class="st">"output/sankey.html"</span>, <span class="at">file =</span> <span class="st">"output/sankey_arg.png"</span>, <span class="at">vwidth =</span> <span class="dv">900</span>, <span class="at">vheight =</span> <span class="dv">900</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="generating-a-sankey-diagram-using-ggsankey." class="level2">
<h2 class="anchored" data-anchor-id="generating-a-sankey-diagram-using-ggsankey.">Generating a Sankey diagram using ggsankey.</h2>
<p>In this example I used the package <code>ggsankey</code> to generate a Sankey diagram. Visualising the beta-lactam resistance genes was not as straight forward as the other resistance genes in the Sankey diagram above. There are multiple enzyme types, which all have a large number of variants. Certain variants have a wider range of resistance compared with their parent types, which adds in more complexity. For this reason I started with the genes on the left hand side.</p>
<section id="packages-1" class="level3">
<h3 class="anchored" data-anchor-id="packages-1">Packages</h3>
<p>Install the packages ggsankey, showtext and sysfonts, glue, patchwok and ggtext then load the libraries</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#remotes::install_github("davidsjoberg/ggsankey")</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggsankey)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(showtext)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sysfonts)</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtext)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="read-in-and-prepare-data-1" class="level3">
<h3 class="anchored" data-anchor-id="read-in-and-prepare-data-1">Read in and prepare data</h3>
<p>Next read in the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/betalactams_gsankey.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>My csv file contains three columns: “gene”, “mechanism”, “subclass”. This data frame is set up differently to the dataframe used for the first Sankey diagram in that each column will be a column in the Sankey diagram. The first column “gene” is the name of each resistance gene. The second column is the mechanism or type of enzyme. Instead of having “enzymatic inactivation” I divided the beta-lactamases into “Penicillinases”, “Oxcillinases”, “IRT” (inhibitor resistant TEM), “ESBL”, “AmpC”, and “Carbapenemases”. In addition, I included the mechanisms “reduced permeability” and “efflux pump”. The third column is the antibiotic subclass. Given most genes confer resistance to multiple sub classes there are multiple rows for each gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Structure of my dataframe</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>spc_tbl_ [121 × 3] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ gene     : chr [1:121] "ompC*" "ompC*" "ompC*" "ompC*" ...
 $ mechanism: chr [1:121] "Reduced permeability" "Reduced permeability" "Reduced permeability" "Reduced permeability" ...
 $ subclass : chr [1:121] "NS penicillins" "ES penicillins" "1G cephalosporins" "3G cephalosporins" ...
 - attr(*, "spec")=
  .. cols(
  ..   gene = col_character(),
  ..   mechanism = col_character(),
  ..   subclass = col_character()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#First six lines of my data frame</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  gene  mechanism            subclass         
  &lt;chr&gt; &lt;chr&gt;                &lt;chr&gt;            
1 ompC* Reduced permeability NS penicillins   
2 ompC* Reduced permeability ES penicillins   
3 ompC* Reduced permeability 1G cephalosporins
4 ompC* Reduced permeability 3G cephalosporins
5 ompC* Reduced permeability Carbapenems      
6 ompF* Reduced permeability NS penicillins   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Last six lines of my data frame</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  gene     mechanism      subclass         
  &lt;chr&gt;    &lt;chr&gt;          &lt;chr&gt;            
1 blaNDM-1 Carbapenemases Carbapenems      
2 blaIMP-4 Carbapenemases 1G cephalosporins
3 blaIMP-4 Carbapenemases 2G cephalosporins
4 blaIMP-4 Carbapenemases 3G cephalosporins
5 blaIMP-4 Carbapenemases 4G cephalosporins
6 blaIMP-4 Carbapenemases Carbapenems      </code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">#The unique mechanisms / enzyme types</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(df<span class="sc">$</span>mechanism)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Reduced permeability" "Efflux pump"          "Penicillinases"      
[4] "Oxcillinases"         "IRT"                  "ESBL"                
[7] "AmpC"                 "Carbapenemases"      </code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co">#The unique subclasses of antibiotics</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(df<span class="sc">$</span>subclass)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "NS penicillins"    "ES penicillins"    "1G cephalosporins"
[4] "3G cephalosporins" "Carbapenems"       "4G cephalosporins"
[7] "Monobactams"       "2G cephalosporins"</code></pre>
</div>
</div>
<p>Next, the data was converted to a long format using a function <code>make_long()</code> from the <code>ggsankey</code> package. This long format contains columns that define how the links flow through the diagram:</p>
<ul>
<li><p><code>x</code>: the name of the current column (called a stage) on the x axis.</p></li>
<li><p><code>next_x</code>: the name of the next stage (NA for the final stage which is why the subclass rows have NA values).</p></li>
<li><p><code>node</code>: the value at the starting node.</p></li>
<li><p><code>next_node</code>: the value at the connecting node.</p></li>
</ul>
<p>The following code was adapted from the <a href="https://r-graph-gallery.com/package/ggsankey.html">r-graph-gallery</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>df_long <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">make_long</span>(gene, mechanism, subclass)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_long)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  x         node                 next_x    next_node           
  &lt;fct&gt;     &lt;chr&gt;                &lt;fct&gt;     &lt;chr&gt;               
1 gene      ompC*                mechanism Reduced permeability
2 mechanism Reduced permeability subclass  NS penicillins      
3 subclass  NS penicillins       &lt;NA&gt;      &lt;NA&gt;                
4 gene      ompC*                mechanism Reduced permeability
5 mechanism Reduced permeability subclass  ES penicillins      
6 subclass  ES penicillins       &lt;NA&gt;      &lt;NA&gt;                </code></pre>
</div>
</div>
<p>Here is a first look at how the data looks when it is plotted.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_long, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">next_x =</span> next_x, <span class="at">node =</span> node, <span class="at">next_node =</span> next_node)) <span class="sc">+</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/ggsankey1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here is the same plot with the links and nodes coloured by node.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_long, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">next_x =</span> next_x, <span class="at">node =</span> node, <span class="at">next_node =</span> next_node,</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fill =</span> <span class="fu">factor</span>(node))) <span class="sc">+</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey</span>() <span class="sc">+</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_discrete</span>(<span class="at">drop =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/ggsankey2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>There is a lot to tidy up. First I removed the grey background, the grid, and x and y axis labels. I find that using <code>theme_void()</code> is the easiest way to do this. I also removed the legend using <code>show.legend = FALSE</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_long, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">next_x =</span> next_x, <span class="at">node =</span> node, <span class="at">next_node =</span> next_node,</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fill =</span> <span class="fu">factor</span>(node))) <span class="sc">+</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey</span>(<span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_discrete</span>(<span class="at">drop =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/ggsankey3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Next I added node labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_long, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">next_x =</span> next_x, <span class="at">node =</span> node, <span class="at">next_node =</span> next_node,</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fill =</span> <span class="fu">factor</span>(node))) <span class="sc">+</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey</span>(<span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> node)) <span class="sc">+</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_discrete</span>(<span class="at">drop =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/ggsankey4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>I wanted to change the colours of the links so that each gene would be coloured according to its mechanism using a more visually appealing colour palette compared with the default colours. First an extra column was added to the long dataframe to group the genes and mechanism rows by mechanism.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>mechanism_lookup <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">select</span>(gene, mechanism) <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">rename</span>(<span class="at">node =</span> gene) <span class="sc">%&gt;%</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">bind_rows</span>(</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>                df <span class="sc">%&gt;%</span> </span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">distinct</span>(mechanism) <span class="sc">%&gt;%</span> </span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">rename</span>(<span class="at">node =</span> mechanism) <span class="sc">%&gt;%</span> </span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">mutate</span>(<span class="at">mechanism =</span> node)) <span class="sc">%&gt;%</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">distinct</span>()</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Add mechanism info to the long data</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>df_long_coloured <span class="ot">&lt;-</span> df_long <span class="sc">%&gt;%</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>        <span class="fu">left_join</span>(mechanism_lookup, <span class="at">by =</span> <span class="st">"node"</span>)</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_long_coloured)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  x         node                 next_x    next_node            mechanism       
  &lt;fct&gt;     &lt;chr&gt;                &lt;fct&gt;     &lt;chr&gt;                &lt;chr&gt;           
1 gene      ompC*                mechanism Reduced permeability Reduced permeab…
2 mechanism Reduced permeability subclass  NS penicillins       Reduced permeab…
3 subclass  NS penicillins       &lt;NA&gt;      &lt;NA&gt;                 &lt;NA&gt;            
4 gene      ompC*                mechanism Reduced permeability Reduced permeab…
5 mechanism Reduced permeability subclass  ES penicillins       Reduced permeab…
6 subclass  ES penicillins       &lt;NA&gt;      &lt;NA&gt;                 &lt;NA&gt;            </code></pre>
</div>
</div>
<p>I selected my colours using the <a href="https://r-graph-gallery.com/color-palette-finder">r-graph-gallery color-palette-finder</a>. Shades of blue from the <code>seeblau</code> palette were used for the “Reduced permeability” and “Efflux pump” mechanisms. Shades of yellow, orange, and red from the <code>amber_material</code> palette were used for the beta-lactamase enzyme types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co">#The two palettes that I selected my colours from:</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">paletteer_d</span>(<span class="st">"unikn::pal_seeblau"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;colors&gt;
#CCEEF9FF #A6E1F4FF #59C7EBFF #00A9E0FF #008ECEFF </code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paletteer_d</span>(<span class="st">"ggsci::amber_material"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;colors&gt;
#FFF8E1FF #FFECB3FF #FFE082FF #FFD54FFF #FFCA28FF #FFC107FF #FFB300FF #FFA000FF #FF8F00FF #FF6F00FF </code></pre>
</div>
</div>
<p>I defined my colour palette. All the antibiotic classes were defined as “other”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>my_colour <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Reduced permeability"</span> <span class="ot">=</span> <span class="st">"#A6E1F4FF"</span>,      </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Efflux pump"</span> <span class="ot">=</span> <span class="st">"#0F7BA2FF"</span>,         </span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Penicillinases"</span> <span class="ot">=</span> <span class="st">"#FFECB3FF"</span>,</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Oxcillinases"</span> <span class="ot">=</span> <span class="st">"#FFD54FFF"</span>, </span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"IRT"</span> <span class="ot">=</span> <span class="st">"#FFC107FF"</span>, </span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"ESBL"</span> <span class="ot">=</span> <span class="st">"#FFA000FF"</span>,</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"AmpC"</span> <span class="ot">=</span> <span class="st">"#FF6F00FF"</span>, </span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Carbapenemases"</span> <span class="ot">=</span> <span class="st">"#DA291CFF"</span>,</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Other"</span> <span class="ot">=</span> <span class="st">"#B3B7B8FF"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>I filled the links by the new mechanism variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>        df_long_coloured, </span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">next_x =</span> next_x, <span class="at">node =</span> node, <span class="at">next_node =</span> next_node,</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fill =</span> mechanism)) <span class="sc">+</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey</span>(<span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> node)) <span class="sc">+</span></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_discrete</span>(<span class="at">drop =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Although I now have the same colour for each mechanism, my colour palette has not been used therefore, <code>scale_fill_discrete()</code> was changed to <code>scale_fill_manual()</code>. The colours on the links were very bright so to make them slightly transparent I used the argument <code>flow.alpha</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>        df_long_coloured, </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">next_x =</span> next_x, <span class="at">node =</span> node, <span class="at">next_node =</span> next_node,</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fill =</span> mechanism)) <span class="sc">+</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey</span>(<span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> node)) <span class="sc">+</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colour) <span class="sc">+</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Next I left justified the first column labels and right justified the second column labels. This code was adapted from this <a href="https://stackoverflow.com/questions/78240951/adjust-labels-on-individual-nodes-in-sankey-diagram-using-ggsankey">stack overflow post</a>. Here the function <code>stage()</code> was used to control the stage at which the aesthetics should be mapped.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>        df_long_coloured, </span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">next_x =</span> next_x,</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">node =</span> node, <span class="at">next_node =</span> next_node,</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">fill =</span> mechanism, <span class="at">label =</span> node)) <span class="sc">+</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey</span>(<span class="at">flow.alpha =</span> <span class="fl">0.7</span>,</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey_text</span>(</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">x =</span> <span class="fu">stage</span>(x, <span class="at">after_stat =</span> x <span class="sc">+</span> <span class="fl">0.1</span> <span class="sc">*</span> <span class="fu">case_when</span>(</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>                                                  x <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span>,    </span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>                                                  x <span class="sc">==</span> <span class="dv">3</span> <span class="sc">~</span> <span class="dv">1</span>,     </span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">.default =</span> <span class="dv">0</span>)),</span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a>                    <span class="at">hjust =</span> <span class="fu">case_when</span>(x <span class="sc">==</span> <span class="st">"gene"</span> <span class="sc">~</span> <span class="dv">1</span>,        </span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a>                            x <span class="sc">==</span> <span class="st">"subclass"</span> <span class="sc">~</span> <span class="dv">0</span>, </span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a>                            <span class="at">.default =</span> <span class="fl">0.5</span>))) <span class="sc">+</span></span>
<span id="cb50-17"><a href="#cb50-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb50-18"><a href="#cb50-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colour)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Next I reordered the nodes. See this <a href="https://github.com/davidsjoberg/ggsankey/issues/7">link</a> for more explanation on how to reorder nodes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>df_long_coloured2 <span class="ot">&lt;-</span> df_long_coloured</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>df_long_coloured2<span class="sc">$</span>node <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  df_long_coloured<span class="sc">$</span>node, </span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">rev</span>(<span class="fu">c</span>(</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ompC*"</span>, <span class="st">"Reduced permeability"</span>, <span class="st">"ompF*"</span>, </span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"acrAB-tolC*"</span>, <span class="st">"Efflux pump"</span>, </span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"blaTEM-1"</span>, <span class="st">"Penicillinases"</span>, <span class="st">"blaSHV-1"</span>, </span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"blaOXA-1"</span>, <span class="st">"Oxcillinases"</span>, </span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"blaTEM-30"</span>, <span class="st">"IRT"</span>, </span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"blaTEM-10"</span>, <span class="st">"ESBL"</span>, <span class="st">"blaSHV-12"</span>, <span class="st">"blaCTX-M-15"</span>, </span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"blaGES-3"</span>, <span class="st">"blaVEB-1"</span>, <span class="st">"blaROB-1"</span>, <span class="st">"blaTLA-1"</span>, </span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ampC*"</span>, <span class="st">"AmpC"</span>, </span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"blaCMY-2"</span>, <span class="st">"blaDHA-1"</span>, <span class="st">"blaFOX-1"</span>, <span class="st">"blaACC-4"</span>, </span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">"blaKPC-18"</span>, <span class="st">"Carbapenemases"</span>, </span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"blaOXA-48"</span>, <span class="st">"blaNDM-1"</span>, <span class="st">"blaIMP-4"</span>,</span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"NS penicillins"</span>, <span class="st">"ES penicillins"</span>, <span class="st">"Monobactams"</span>, </span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1G cephalosporins"</span>, <span class="st">"2G cephalosporins"</span>, </span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"3G cephalosporins"</span>, <span class="st">"4G cephalosporins"</span>, <span class="st">"Carbapenems"</span>))</span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a>df_long_coloured2<span class="sc">$</span>next_node <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>  df_long_coloured<span class="sc">$</span>next_node, </span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">rev</span>(<span class="fu">c</span>(<span class="st">"Reduced permeability"</span>, <span class="st">"Efflux pump"</span>, <span class="st">"Penicillinases"</span>, </span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Oxcillinases"</span>, <span class="st">"IRT"</span>, <span class="st">"ESBL"</span>, <span class="st">"AmpC"</span>, <span class="st">"Carbapenemases"</span>,</span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a>    <span class="st">"NS penicillins"</span>, <span class="st">"ES penicillins"</span>,<span class="st">"Monobactams"</span>,<span class="st">"1G cephalosporins"</span>,</span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Carbapenems"</span>,<span class="st">"2G cephalosporins"</span>,<span class="st">"3G cephalosporins"</span>, <span class="st">"4G cephalosporins"</span>))</span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The plot with the reordered nodes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_long_coloured2, <span class="fu">aes</span>(</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> x, <span class="at">next_x =</span> next_x,</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">node =</span> node, <span class="at">next_node =</span> next_node,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">fill =</span> mechanism, <span class="at">label =</span> node)) <span class="sc">+</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey</span>(<span class="at">flow.alpha =</span> <span class="fl">0.7</span>,</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_sankey_text</span>(<span class="fu">aes</span>(</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">x =</span> <span class="fu">stage</span>(x, <span class="at">after_stat =</span> x <span class="sc">+</span> <span class="fl">0.1</span> <span class="sc">*</span><span class="fu">case_when</span>(</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>                                                  x <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span>,    </span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>                                                  x <span class="sc">==</span> <span class="dv">3</span> <span class="sc">~</span> <span class="dv">1</span>,     </span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">.default =</span> <span class="dv">0</span>)),</span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>                    <span class="at">hjust =</span> <span class="fu">case_when</span>(x <span class="sc">==</span> <span class="st">"gene"</span> <span class="sc">~</span> <span class="dv">1</span>,        </span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>                            x <span class="sc">==</span> <span class="st">"subclass"</span> <span class="sc">~</span> <span class="dv">0</span>, </span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>                            <span class="at">.default =</span> <span class="fl">0.5</span>))) <span class="sc">+</span></span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colour)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>I italicised the gene names and changed the mechanisms to a bold style using plotmath syntax. First I added another column, which will be used for plotmath syntax, called label_pms. For example, to generate <code>x</code> in italic font use <code>italic(x)</code>. Go to this <a href="https://www.rdocumentation.org/packages/grDevices/versions/3.6.2/topics/plotmath">link</a> to read more on plotmath syntax. The <code>paste()</code> function was used to join strings together with no separator. I joined either the string <code>"italic('"</code>, <code>"bold('"</code>, or <code>"plain('"</code> with the node value (so the gene name), and the string <code>"')"</code>. Then I changed those labels with a <em>bla</em> gene so that the enzyme type was written in subscript. The function <code>str_replace_all()</code> was used to replace any text with the literal string <code>italic\\(</code> followed by <code>bla</code> with the pattern of three word characters<code>\\w{3}</code>, a literal hyphen <code>-</code>, one word character <code>\\w</code> followed by one or more digits <code>\\d</code> or (<code>|</code>) the pattern of three word characters, hyphen, multiple digits.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Add a new column called label_pms with the plotmath syntax for font styling</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>df_long_coloured3 <span class="ot">&lt;-</span> df_long_coloured2 <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">label_pms =</span> <span class="fu">case_when</span>(</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>    x <span class="sc">==</span> <span class="st">"gene"</span> <span class="sc">~</span> <span class="fu">paste0</span>(<span class="st">"italic('"</span>, node, <span class="st">"')"</span>),          </span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>    x <span class="sc">==</span> <span class="st">"mechanism"</span> <span class="sc">~</span> <span class="fu">paste0</span>(<span class="st">"bold('"</span>, node, <span class="st">"')"</span>),       </span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>    x <span class="sc">==</span> <span class="st">"subclass"</span> <span class="sc">~</span> <span class="fu">paste0</span>(<span class="st">"plain('"</span>, node, <span class="st">"')"</span>)        </span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Change the bla gene labels to subscript after bla</span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>df_long_coloured3 <span class="ot">&lt;-</span> df_long_coloured3 <span class="sc">%&gt;%</span></span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">label_pms =</span> <span class="fu">str_replace_all</span>(label_pms,                                <span class="st">"italic</span><span class="sc">\\</span><span class="st">('(bla)(</span><span class="sc">\\</span><span class="st">w{3}-</span><span class="sc">\\</span><span class="st">w-</span><span class="sc">\\</span><span class="st">d+|</span><span class="sc">\\</span><span class="st">w{3}-</span><span class="sc">\\</span><span class="st">d+)'</span><span class="sc">\\</span><span class="st">)"</span>, </span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"italic('</span><span class="sc">\\</span><span class="st">1')[</span><span class="sc">\\</span><span class="st">2]"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that when you change the order of your nodes, you also need to set the levels for your labels so that they match the same order as your nodes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>df_long_coloured3<span class="sc">$</span>label_pms <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  df_long_coloured3<span class="sc">$</span>label_pms,</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">rev</span>(<span class="fu">c</span>(</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('ompC*')"</span>,</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bold('Reduced permeability')"</span>,</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('ompF*')"</span>,</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('acrAB-tolC*')"</span>,</span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bold('Efflux pump')"</span>,</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[TEM-1]"</span>,</span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bold('Penicillinases')"</span>,</span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[SHV-1]"</span>,</span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[OXA-1]"</span>,</span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bold('Oxcillinases')"</span>,</span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[TEM-30]"</span>,</span>
<span id="cb54-15"><a href="#cb54-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bold('IRT')"</span>,</span>
<span id="cb54-16"><a href="#cb54-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[TEM-10]"</span>,</span>
<span id="cb54-17"><a href="#cb54-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bold('ESBL')"</span>,</span>
<span id="cb54-18"><a href="#cb54-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[SHV-12]"</span>,</span>
<span id="cb54-19"><a href="#cb54-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[CTX-M-15]"</span>,</span>
<span id="cb54-20"><a href="#cb54-20" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[GES-3]"</span>,</span>
<span id="cb54-21"><a href="#cb54-21" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[VEB-1]"</span>,</span>
<span id="cb54-22"><a href="#cb54-22" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[ROB-1]"</span>,</span>
<span id="cb54-23"><a href="#cb54-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[TLA-1]"</span>,</span>
<span id="cb54-24"><a href="#cb54-24" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('ampC*')"</span>,</span>
<span id="cb54-25"><a href="#cb54-25" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bold('AmpC')"</span>,</span>
<span id="cb54-26"><a href="#cb54-26" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[CMY-2]"</span>,</span>
<span id="cb54-27"><a href="#cb54-27" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[DHA-1]"</span>,</span>
<span id="cb54-28"><a href="#cb54-28" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[FOX-1]"</span>,</span>
<span id="cb54-29"><a href="#cb54-29" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[ACC-4]"</span>,</span>
<span id="cb54-30"><a href="#cb54-30" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[KPC-18]"</span>,</span>
<span id="cb54-31"><a href="#cb54-31" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bold('Carbapenemases')"</span>,</span>
<span id="cb54-32"><a href="#cb54-32" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[OXA-48]"</span>,</span>
<span id="cb54-33"><a href="#cb54-33" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[NDM-1]"</span>,</span>
<span id="cb54-34"><a href="#cb54-34" aria-hidden="true" tabindex="-1"></a>    <span class="st">"italic('bla')[IMP-4]"</span>,</span>
<span id="cb54-35"><a href="#cb54-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Now the antibiotic subclasses in your desired order:</span></span>
<span id="cb54-36"><a href="#cb54-36" aria-hidden="true" tabindex="-1"></a>    <span class="st">"plain('NS penicillins')"</span>,</span>
<span id="cb54-37"><a href="#cb54-37" aria-hidden="true" tabindex="-1"></a>    <span class="st">"plain('ES penicillins')"</span>,</span>
<span id="cb54-38"><a href="#cb54-38" aria-hidden="true" tabindex="-1"></a>    <span class="st">"plain('Monobactams')"</span>,</span>
<span id="cb54-39"><a href="#cb54-39" aria-hidden="true" tabindex="-1"></a>    <span class="st">"plain('1G cephalosporins')"</span>,</span>
<span id="cb54-40"><a href="#cb54-40" aria-hidden="true" tabindex="-1"></a>    <span class="st">"plain('2G cephalosporins')"</span>,</span>
<span id="cb54-41"><a href="#cb54-41" aria-hidden="true" tabindex="-1"></a>    <span class="st">"plain('3G cephalosporins')"</span>,</span>
<span id="cb54-42"><a href="#cb54-42" aria-hidden="true" tabindex="-1"></a>    <span class="st">"plain('4G cephalosporins')"</span>,</span>
<span id="cb54-43"><a href="#cb54-43" aria-hidden="true" tabindex="-1"></a>    <span class="st">"plain('Carbapenems')"</span></span>
<span id="cb54-44"><a href="#cb54-44" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb54-45"><a href="#cb54-45" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When calling the <code>ggplot()</code> function I then added the arguments <code>label = label_pms</code> and <code>parse = TRUE</code>, to parse it as plotmath syntax.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>ps <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_long_coloured3, <span class="fu">aes</span>(</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> x, <span class="at">next_x =</span> next_x,</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">node =</span> node, <span class="at">next_node =</span> next_node,</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> mechanism, <span class="at">label_pms =</span> node)) <span class="sc">+</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sankey</span>(<span class="at">flow.alpha =</span> <span class="fl">0.7</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sankey_text</span>(</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> label_pms,</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="fu">stage</span>(x, <span class="at">after_stat =</span> x <span class="sc">+</span> <span class="fl">0.1</span> <span class="sc">*</span></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>          <span class="fu">case_when</span>(</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>            x <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span>,</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>            x <span class="sc">==</span> <span class="dv">3</span> <span class="sc">~</span> <span class="dv">1</span>,</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">.default =</span> <span class="dv">0</span>)),</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">hjust =</span> <span class="fu">case_when</span>(x <span class="sc">==</span> <span class="st">"gene"</span> <span class="sc">~</span> <span class="dv">1</span>,</span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>                          x <span class="sc">==</span> <span class="st">"subclass"</span> <span class="sc">~</span> <span class="dv">0</span>,</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a>                          <span class="at">.default =</span> <span class="fl">0.5</span>)),</span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">parse =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb55-18"><a href="#cb55-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colour)</span>
<span id="cb55-19"><a href="#cb55-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-20"><a href="#cb55-20" aria-hidden="true" tabindex="-1"></a>ps</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Lastly I added a caption. For adding captions, titles and customising ggplots I would recommend Nicola Rennie’s book ’The Art of Data Visualization with ggplot2”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>caption1 <span class="ot">&lt;-</span> <span class="st">"Beta-lactam resistance in &lt;i&gt;E. coli&lt;/i&gt; predominantly occurs through the acquisition of genes encoding beta-lactamases (shades of yellow, orange, and red, according to enzyme group), but can also occur through mutations in genes encoding outer membrane proteins (light blue) or efflux pumps (dark blue); genes with mutations are marked with an asterisk. This Sankey diagram illustrates how beta-lactam resistance genes can be grouped by mechanism (reduced permeability, efflux pump and enzymatic inactivation by different beta-lactamases) and the subclasses of beta-lactams (NS - narrow spectrum and ES - extended spectrum penicllins; monobactams; 1G - first generation, 2G - second generation, 3G - third generation, 4G - fourth generation; and carbapenems) to which they confer resistance. The beta-lactamase enzymes have been grouped into penicillinases, oxacillinases, IRT (inhibitor resistant TEM), ESBL, AmpC, and carbapenemases. One representative gene variant is shown. However, resistance profiles can vary between different variants of the same enzyme and between different enzymes from the same group."</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>psc <span class="ot">&lt;-</span> ps <span class="sc">+</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">caption =</span> caption1) <span class="sc">+</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption.position =</span> <span class="st">"plot"</span>, </span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.caption =</span> <span class="fu">element_textbox_simple</span>(</span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">hjust =</span> <span class="fl">0.5</span>, </span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="dv">40</span>, <span class="at">r =</span> <span class="dv">5</span>, <span class="at">b =</span> <span class="dv">5</span>, <span class="at">l =</span> <span class="dv">5</span>)))</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>psc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/caption1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>To attribute this work please cite:</p>
<p>Burgess, Sara. 2025. “Generating and polishing Sankey diagrams in R.” 14 November, 2025. https://sburgess2.github.io/sankey_amr/</p>



</section>
</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-sidjabat_interspecies_2014" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Sidjabat, H. E., Heney, C., George, N. M., Nimmo, G. R. &amp; Paterson, D. L. <a href="https://doi.org/10.1128/JCM.01491-14">Interspecies <span>Transfer</span> of <em>bla</em><span class="math inline">\(_{\text{IMP-4}}\)</span> in a <span>Patient</span> with <span>Prolonged</span> <span>Colonization</span> by <span>IMP</span>-4-<span>Producing</span> <span>Enterobacteriaceae</span></a>. <em>Journal of Clinical Microbiology</em> <strong>52</strong>, 3816–3818 (2014).</div>
</div>
<div id="ref-mmatli_emerging_2020" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Mmatli, M., Mbelle, N. M., Maningi, N. E. &amp; Osei Sekyere, J. <a href="https://doi.org/10.1128/mSystems.00783-20">Emerging <span>Transcriptional</span> and <span>Genomic</span> <span>Mechanisms</span> <span>Mediating</span> <span>Carbapenem</span> and <span>Polymyxin</span> <span>Resistance</span> in <em>enterobacteriaceae</em>: A <span>Systematic</span> <span>Review</span> of <span>Current</span> <span>Reports</span></a>. <em>mSystems</em> <strong>5</strong>, e00783–20 (2020).</div>
</div>
<div id="ref-vourli_novel_2004" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Vourli, S. <a href="https://doi.org/10.1016/j.femsle.2004.03.028">Novel <span>GES</span>/<span>IBC</span> extended-spectrum β-lactamase variants with carbapenemase activity in clinical enterobacteria</a>. <em>FEMS Microbiology Letters</em> <strong>234</strong>, 209–213 (2004).</div>
</div>
<div id="ref-dolatyar_dehkharghani_clonal_2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Dolatyar Dehkharghani, A., Haghighat, S., Rahnamaye Farzami, M., Rahbar, M. &amp; Douraghi, M. <a href="https://doi.org/10.3389/fcimb.2021.560622">Clonal <span>Relationship</span> and <span>Resistance</span> <span>Profiles</span> <span>Among</span> <span>ESBL</span>-<span>Producing</span> <em><span>Escherichia</span> coli</em></a>. <em>Frontiers in Cellular and Infection Microbiology</em> <strong>11</strong>, 560622 (2021).</div>
</div>
<div id="ref-papagiannitsis_plasmid-encoded_2007" class="csl-entry" role="listitem">
<div class="csl-left-margin">5. </div><div class="csl-right-inline">Papagiannitsis, C. C., Tzouvelekis, L. S., Tzelepi, E. &amp; Miriagou, V. <a href="https://doi.org/10.1128/AAC.00389-07">Plasmid-<span>Encoded</span> <span>ACC</span>-4, an <span>Extended</span>-<span>Spectrum</span> <span>Cephalosporinase</span> <span>Variant</span> from <em><span>Escherichia</span> coli</em></a>. <em>Antimicrobial Agents and Chemotherapy</em> <strong>51</strong>, 3763–3767 (2007).</div>
</div>
<div id="ref-brinas_-lactamases_2002" class="csl-entry" role="listitem">
<div class="csl-left-margin">6. </div><div class="csl-right-inline">Briñas, L., Zarazaga, M., Sáenz, Y., Ruiz-Larrea, F. &amp; Torres, C. <a href="https://doi.org/10.1128/AAC.46.10.3156-3163.2002">Β-<span>Lactamases</span> in <span>Ampicillin</span>-<span>Resistant</span> <em><span>Escherichia</span> coli</em> <span>Isolates</span> from <span>Foods</span>, <span>Humans</span>, and <span>Healthy</span> <span>Animals</span></a>. <em>Antimicrobial Agents and Chemotherapy</em> <strong>46</strong>, 3156–3163 (2002).</div>
</div>
<div id="ref-hao_genotypic_2019" class="csl-entry" role="listitem">
<div class="csl-left-margin">7. </div><div class="csl-right-inline">Hao, Y. <em>et al.</em> <a href="https://doi.org/10.3389/fmicb.2019.00788">Genotypic and <span>Phenotypic</span> <span>Characterization</span> of <span>Clinical</span> <em><span>Escherichia</span> coli</em> <span>Sequence</span> <span>Type</span> 405 <span>Carrying</span> <span>IncN2</span> <span>Plasmid</span> <span>Harboring</span> <em>bla</em><span class="math inline">\(_{\text{NDM-1}}\)</span></a>. <em>Frontiers in Microbiology</em> <strong>10</strong>, 788 (2019).</div>
</div>
<div id="ref-kazmierczak_global_2016" class="csl-entry" role="listitem">
<div class="csl-left-margin">8. </div><div class="csl-right-inline">Kazmierczak, K. M. <em>et al.</em> <a href="https://doi.org/10.1128/AAC.00107-16">Global <span>Dissemination</span> of <em>bla</em><span class="math inline">\(_{\text{KPC}}\)</span> into <span>Bacterial</span> <span>Species</span> beyond <em><span>Klebsiella</span> pneumoniae</em> and <em><span>In</span> <span>Vitro</span></em> <span>Susceptibility</span> to <span>Ceftazidime</span>-<span>Avibactam</span> and <span>Aztreonam</span>-<span>Avibactam</span></a>. <em>Antimicrobial Agents and Chemotherapy</em> <strong>60</strong>, 4490–4500 (2016).</div>
</div>
<div id="ref-vale_detection_2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">9. </div><div class="csl-right-inline">Vale, A. P., Shubin, L., Cummins, J., Leonard, F. C. &amp; Barry, G. <a href="https://doi.org/10.3389/fvets.2021.583759">Detection of <em>bla</em><span class="math inline">\(_{\text{OXA-1}}\)</span>, <em>bla</em><span class="math inline">\(_{\text{TEM-1}}\)</span>, and <span>Virulence</span> <span>Factors</span> in <em><span>E</span>. Coli</em> <span>Isolated</span> <span>From</span> <span>Seals</span></a>. <em>Frontiers in Veterinary Science</em> <strong>8</strong>, 583759 (2021).</div>
</div>
<div id="ref-liakopoulos_review_2016" class="csl-entry" role="listitem">
<div class="csl-left-margin">10. </div><div class="csl-right-inline">Liakopoulos, A., Mevius, D. &amp; Ceccarelli, D. <a href="https://doi.org/10.3389/fmicb.2016.01374">A <span>Review</span> of <span>SHV</span> <span>Extended</span>-<span>Spectrum</span> β-<span>Lactamases</span>: <span>Neglected</span> <span>Yet</span> <span>Ubiquitous</span></a>. <em>Frontiers in Microbiology</em> <strong>7</strong>, (2016).</div>
</div>
<div id="ref-zhang_vitro_2020" class="csl-entry" role="listitem">
<div class="csl-left-margin">11. </div><div class="csl-right-inline">Zhang, B. <em>et al.</em> <a href="https://doi.org/10.1016/j.ijid.2020.05.075">In vitro activity of aztreonam–avibactam against metallo-β-lactamase-producing <span>Enterobacteriaceae</span>—<span>A</span> multicenter study in <span>China</span></a>. <em>International Journal of Infectious Diseases</em> <strong>97</strong>, 11–18 (2020).</div>
</div>
<div id="ref-bradford_extended-spectrum_2001" class="csl-entry" role="listitem">
<div class="csl-left-margin">12. </div><div class="csl-right-inline">Bradford, P. A. <a href="https://doi.org/10.1128/CMR.14.4.933-951.2001">Extended-<span>Spectrum</span> β-<span>Lactamases</span> in the 21st <span>Century</span>: <span>Characterization</span>, <span>Epidemiology</span>, and <span>Detection</span> of <span>This</span> <span>Important</span> <span>Resistance</span> <span>Threat</span></a>. <em>Clinical Microbiology Reviews</em> <strong>14</strong>, 933–951 (2001).</div>
</div>
<div id="ref-kang_genetic_2022" class="csl-entry" role="listitem">
<div class="csl-left-margin">13. </div><div class="csl-right-inline">Kang, H. J., Lim, S.-K. &amp; Lee, Y. J. <a href="https://doi.org/10.3389/fvets.2022.1055320">Genetic characterization of third- or fourth-generation cephalosporin-resistant avian pathogenic <span>Escherichia</span> coli isolated from broilers</a>. <em>Frontiers in Veterinary Science</em> <strong>9</strong>, 1055320 (2022).</div>
</div>
<div id="ref-jacoby_ampc_2009" class="csl-entry" role="listitem">
<div class="csl-left-margin">14. </div><div class="csl-right-inline">Jacoby, G. A. <a href="https://doi.org/10.1128/CMR.00036-08"><span>AmpC</span> β-<span>Lactamases</span></a>. <em>Clinical Microbiology Reviews</em> <strong>22</strong>, 161–182 (2009).</div>
</div>
<div id="ref-bush_epidemiology_2020" class="csl-entry" role="listitem">
<div class="csl-left-margin">15. </div><div class="csl-right-inline">Bush, K. &amp; Bradford, P. A. <a href="https://doi.org/10.1128/CMR.00047-19">Epidemiology of β-<span>Lactamase</span>-<span>Producing</span> <span>Pathogens</span></a>. <em>Clinical Microbiology Reviews</em> <strong>33</strong>, e00047–19 (2020).</div>
</div>
<div id="ref-rennie2025" class="csl-entry" role="listitem">
<div class="csl-left-margin">16. </div><div class="csl-right-inline">Rennie, N. Introducing <span>The</span> <span>Art</span> of <span>Visualization</span> with ggplot2. <a href="https://nrennie.rbind.io/blog/art-of-viz-book/">https://nrennie.rbind.io/blog/art-of-viz-book/</a> (2025).</div>
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>